feat: Encapsulate hardware-specific operations

splasky opened the pull request:
These commit did the following things:

1. Add abstraction layer to support different hardware like `get_key` and `get_device_id`.
2. Rebase onto current develop branch.
3. Update return type inside `conn_http.c` functions.

howjmay commented on hardware_deps/Makefile:
```diff
@@ -0,0 +1,14 @@
+	$(MAKE) -C $(BUILD_PATH)
+	$(CC) -c $(CFLAGS) $(INCLUDES) -MMD -MF device.c.d -o $@ $<
+
+-include device.c.d
```
add new line here

splasky replied:
Should we also remove the `-M` command which help for check dependencies during compiling?

howjmay replied:
I am OK witth taht at this step, but I think we may need this in the future

howjmay commented on hardware_deps/device.h:
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
```
2020

howjmay commented on hardware_deps/sierra_wireless/wp7702/Makefile:
```diff
@@ -0,0 +1,7 @@
+
+hardware_deps.o: hardware_deps.c
+	$(CC) -c $(CFLAGS) $(INCLUDES) $<.d -o $@ $<
+	
```
newline

howjmay commented on hardware_deps/sierra_wireless/wp7702/hardware_deps.h:
```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
```
2020

howjmay commented on hardware_deps/device.c:
```diff
@@ -0,0 +1,21 @@
+
+retcode_t release_device(device_t *d) {
+  free(d);
+  return RET_OK;
```
Do you think we should assign NULL to the pointer of  `*d`?
And it may be good to use `void` instead of `retcode_t`

splasky replied:
Is check the double free problem still good style in 2020? I am not sure.

howjmay replied:
I am free for your decision. Don't worry

howjmay commented on hardware_deps/sierra_wireless/wp7702/hardware_deps.c:
```diff
@@ -0,0 +1,59 @@
+
+  if (pclose(fp) == -1) {
+    perror("close FILE pointer");
+    return RET_OK;
```
It is an error case. We should return error here.

howjmay commented on hardware_deps/sierra_wireless/wp7702/hardware_deps.c:
```diff
@@ -0,0 +1,59 @@
+    return RET_OK;
+  }
+
+  return 0;
```
This one should return `RET_OK`

jserv commented on hardware_deps/Makefile:
```diff
@@ -0,0 +1,14 @@
+
+ifeq ($(SIERRA_WP7702), y)
+	BUILD_PATH = sierra_wireless/wp7702
+	HARDWARE_CFLAG = CFLAGS -DSIERRA_WP7702
```
Change `SIERRA_WP7702` to `HW_SIERRA_WP7702`.

jserv replied:
Change `HARDWARE_CFLAG` to `HW_CFLAGS`.

jserv commented:
Change ` hardware_deps` to `hal`, which means hardware abstraction layer.

jserv commented on utils/uart_utils.h:
```diff
@@ -10,4 +10,4 @@
 void uart_write(const int fd, char *cmd);
-char *uart_read(const int fd);
\ No newline at end of file
+char *uart_read(const int fd);
```
These operations should be encapsulated in HAL.

splasky replied:
New issues #32 for this.

jserv commented:
Let's re-organize the directory structure:
1. Directory `hal`: generic code and interface; We need documentation though.
2. Directory `devices/sierra_wireless/wp7702`: The exact hardware-specific implementation;

jserv commented on hal/device.c:
```diff
@@ -0,0 +1,19 @@
+#include <stdlib.h>
+
+// TODO: binding device using hal not compiler
+device_t *hardware_deps_init(void) {
```
It looks ugly. Can you use something like Linux VFS for registering machine-dependent entries?

jserv commented on hal/device.h:
```diff
@@ -0,0 +1,39 @@
+ * "LICENSE" at the root of this distribution.
+ */
+
+#ifndef DEVICE_H
```
Should be `HAL_DEVICE_INTERFACE_H` or something alike.

jserv commented on hal/devices/sierra_wireless/wp7702/Makefile:
```diff
@@ -0,0 +1,2 @@
+wp7702.o: wp7702.c
```
Let's unify filename. That is, `wp7702.c` should be changed into something like `entry.c` or `impl.c`

jserv commented on hal/devices/sierra_wireless/wp7702/wp7702.h:
```diff
@@ -0,0 +1,25 @@
+/*
```
Similarly, the header should be unified.

howjmay commented on connectivity/conn_http.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * Copyright (C) 2020 BiiLabs Co., Ltd. and Contributors
```
it should be `2019-2020`

howjmay commented on utils/crypto_utils.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2018-2019 BiiLabs Co., Ltd. and Contributors
+ * Copyright (C) 2020 BiiLabs Co., Ltd. and Contributors
```
2019-2020

howjmay commented on utils/defined_error.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * Copyright (C) 2020 BiiLabs Co., Ltd. and Contributors
```
2019-2020

howjmay commented on utils/serializer.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * Copyright (C) 2020 BiiLabs Co., Ltd. and Contributors
```
ditto

howjmay commented on utils/serializer.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * Copyright (C) 2020 BiiLabs Co., Ltd. and Contributors
```
ditto

howjmay commented on utils/tryte_byte_conv.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * Copyright (C) 2020 BiiLabs Co., Ltd. and Contributors
```
dutto

howjmay commented on utils/tryte_byte_conv.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * Copyright (C) 2020 BiiLabs Co., Ltd. and Contributors
```
ditto

howjmay commented on utils/uart_utils.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * Copyright (C) 2020 BiiLabs Co., Ltd. and Contributors
```
ditto

howjmay commented on utils/uart_utils.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * Copyright (C) 2020 BiiLabs Co., Ltd. and Contributors
```
ditto

howjmay commented on main.c:
```diff
@@ -40,10 +40,11 @@ void gen_trytes(uint16_t len, char *out) {
 
-void send_https_msg(char const *const host, char const *const port, char const *const api, char const *const tryte_msg,
-                    char const *const addr) {
+int send_https_msg(char const *const host, char const *const port, char const *const api, char const *const tryte_msg,
```
how about using retcode_t here?

jserv commented on hal/device.h:
```diff
@@ -0,0 +1,36 @@
+ * "LICENSE" at the root of this distribution.
+ */
+
+#ifndef DEVICE_H
```
Should be `HAL_DEVICE_H`

jserv commented on hal/devices/sierra_wireless/wp7702/impl.c:
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
```
Let's move `devices` from `hal`. That is, we have two directories:
* `hal`: the abstract interface which allows high-level access;
* `devices`: the real device manipulations;

jserv commented on devices/sierra_wireless/wp7702/impl.c:
```diff
@@ -0,0 +1,91 @@
+#include <string.h>
+#include <time.h>
+
+// TODO:Remove here after dev merged
```
Improve the comment!
Don't say something like `after dev merged`.

jserv commented on hal/device.h:
```diff
@@ -0,0 +1,36 @@
+struct device_s {
+  void (*destructor)(device_t *);
+  retcode_t (*get_key)(uint8_t *key);                    /**< get device private key */
+  retcode_t (*get_hardware_identifier)(char *device_id); /**< get device id */
```
Change to `get_device_id`.

jserv replied:
We have to encapsulate hardware-specific operations as possible as we can. @HowJMay, I need your input.

jserv commented on hal/device.h:
```diff
@@ -0,0 +1,36 @@
+ */
+typedef struct device_s device_t;
+struct device_s {
+  void (*destructor)(device_t *);
```
Shall we provide `init`?

jserv replied:
How about changing from `destructor` to `fini`?

splasky replied:
> Shall we provide `init`?

Will it be helpful?  How will the object be generated after we improve the `init` inside device_s class?

jserv replied:
If we provide `init`, we can eliminate the assignment for UART device (such as `ttyUSB0`) where we can focus on the logic behind device manipulation.

jserv commented:
By the way, @splasky should create a dedicated issue which describes the details this pull request would concentrate on. Ask @HowJMay for revising.

jserv commented:
The policy is "every pull request is used to close existing issue(s)."

jserv commented on hal/device.h:
```diff
@@ -0,0 +1,37 @@
+ */
+typedef struct device_s device_t;
+struct device_s {
+  void (*ini)(device_t *);            /**< init for device */
```
Typo.

jserv replied:
Should be `init`.

jserv commented on devices/sierra_wireless/wp7702/impl.c:
```diff
@@ -0,0 +1,91 @@
+static retcode_t wp7702_get_hardware_identifier(char *device_id);
+static void wp7702_release(device_t *);
+
+device_t *new_wp7702(void) {
```
Since we are going to implement abstract interface, we should avoid exposing functions such as `new_wp7702`.

jserv replied:
Thinking of the case in Linux VFS, Linux has many file systems but there are only descriptors exposed.

jserv commented:
Don't write git commit messages starting with "Draft."

Commit:
[WIP] Split commits and push to new pull requests

1. Implement hal
2. wp7702 support
3. emulator support
4. Add new logger.h for logging

Closes #32

jserv commented on devices/fake_device/Makefile:
```diff
@@ -0,0 +1,2 @@
+fake_device.o: impl.c uart.c
```
Change it to `emulator`.

jserv commented on devices/fake_device/impl.c:
```diff
@@ -0,0 +1,115 @@
+#define IMSI_LEN 15
+#define AES_BLOCK_SIZE 16
+
+DECLARE_DEVICE(wp7702);
```
Revise device information.

jserv commented on devices/fake_device/uart.c:
```diff
@@ -15,6 +13,16 @@
 #include <unistd.h>
+#include "device.h"
+
+static char *portname = "/dev/ttyHS0";
```
Don't mention real device within emulator/simulator.

jserv commented on tests/test_main.c:
```diff
@@ -1,3 +1,11 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
```
Don't use the name. Change it to `test_driver.c`.

jserv reviewed (changes requested):
Revise the target name and its property.

jserv commented on devices/emulator/impl.c:
```diff
@@ -0,0 +1,91 @@
+
+extern struct uart_operations emulator_uart;
+
+static struct device_operations emulator_ops = {.init = &emulator_init,
```
Avoiding using the form `&emulator_init`. Use `.init = emulator_init` instead. Be aware of consistent indention.

jserv replied:
Ready to resolve?

splasky replied:
Yes, I will do this tomorrow. Code review can also start.

jserv commented on devices/sierra_wireless/wp7702/impl.c:
```diff
@@ -0,0 +1,122 @@
+
+static struct logging_operations wp7702_logger = {.write_log = write_log};
+
+static struct device_type wp7702_device_type = {
```
Move the statement following the callback functions.

jserv commented:
Let's split "emulator" device support out of this pull request, which eases the effort of reviewing.

jserv commented:
Expected procedure for HAL related changes:
1. Proposed common interface for HAL;
2. Re-implement WP7702 specific support with HAL;
3. Write down brief documentation about HAL;
4. (in another pull request) Provide "emulator" or "simulator" device with HAL;

Ask @HowJMay for mentoring.

jserv commented on utils/tryte_msg.h:
```diff
@@ -0,0 +1,23 @@
+#ifndef TRYTE_MSG_H_WJ0SDDW2
+#define TRYTE_MSG_H_WJ0SDDW2
```
Use canonical identifiers.

jserv commented on utils/url.h:
```diff
@@ -0,0 +1,27 @@
+ */
+retcode_t send_https_msg(char const *host, char const *port, char const *api, const char *msg, const int msg_len,
+                         const char *ssl_seed);
+#endif /* end of include guard: UTILS_URL_H */
```
Don't say "end of include guard: UTILS_URL_H."
I hate your stupid editor plugin. Remove it.

splasky commented:
This pull request is too messy and cannot clearly express subject. I will close this pull request and split commits for more easily for reviewing.
